crates/bevy_ecs/src/query/iter.rs

```diff
@@ -849,12 +913,24 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Iterator for QueryIter<'w, 's, D, F>
                 let archetype =
                     // SAFETY: Matched archetype IDs are guaranteed to still exist.
                     unsafe { self.archetypes.get(id.archetype_id).debug_checked_unwrap() };
-                accum =
+                // SAFETY: Matched table IDs are guaranteed to still exist.
+                let table = unsafe { self.tables.get(archetype.table_id()).debug_checked_unwrap() };
+                if table.entity_count() == archetype.len() {
```

Comment:
I think it would be worthwhile to add a comment explaining that this an optimization for hybrid queries where the archetype is actually dense